File: src/ObjectManager/Object.Tes/FilePacks/Records/045-AMMO.Ammo.cs
using OA.Core;

namespace OA.Tes.FilePacks.Records
{
    public class AMMORecord : Record, IHaveEDID, IHaveMODL
    {
        public struct DATAField
        {
            public float Speed;
            public uint Flags;
            public uint Value;
            public float Weight;
            public ushort Damage;

            public DATAField(UnityBinaryReader r, int dataSize)
            {
                Speed = r.ReadLESingle();
                Flags = r.ReadLEUInt32();
                Value = r.ReadLEUInt32();
                Weight = r.ReadLESingle();
                Damage = r.ReadLEUInt16();
            }
        }

        public override string ToString() => $"AMMO: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model
        public STRVField FULL; // Item Name
        public FILEField? ICON; // Male Icon (optional)
        public FMIDField<ENCHRecord>? ENAM; // Enchantment ID (optional)
        public IN16Field? ANAM; // Enchantment points (optional)
        public DATAField DATA; // Ammo Data

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL": FULL = new STRVField(r, dataSize); return true;
                case "ICON": ICON = new FILEField(r, dataSize); return true;
                case "ENAM": ENAM = new FMIDField<ENCHRecord>(r, dataSize); return true;
                case "ANAM": ANAM = new IN16Field(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-AMMO.Ammo.swift
//
//  AMMORecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class AMMORecord: Record, IHaveEDID, IHaveMODL {
    public struct DATAField {
        public let speed: Float
        public let flags: UInt32
        public let value: UInt32
        public let weight: Float
        public let damage: UInt16

        init(_ r: BinaryReader, _ dataSize: Int) {
            speed = r.readLESingle()
            flags = r.readLEUInt32()
            value = r.readLEUInt32()
            weight = r.readLESingle()
            damage = r.readLEUInt16()
        }
    }

    public override var description: String { return "AMMO: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var MODL: MODLGroup? = nil // Model
    public var FULL: STRVField! // Item Name
    public var ICON: FILEField? = nil// Male Icon (optional)
    public var ENAM: FMIDField<ENCHRecord>? = nil // Enchantment ID (optional)
    public var ANAM: IN16Field? = nil // Enchantment points (optional)
    public var DATA: DATAField! // Ammo Data
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "FULL": FULL = r.readSTRV(dataSize)
        case "ICON": ICON = r.readSTRV(dataSize)
        case "ENAM": ENAM = FMIDField<ENCHRecord>(r, dataSize)
        case "ANAM": ANAM = r.readT(dataSize)
        case "DATA": DATA = DATAField(r, dataSize)
        default: return false
        }
        return true
    }
}
